lottery/src/lib.rs

```diff
@@ -0,0 +1,342 @@
+    /// * List of players must not be empty
+    fn get_players(&mut self) {
+        if self.lottery_is_on() && !self.players.is_empty() {
+            msg::reply(Event::Players(self.players.clone()), 0);
+        } else {
+            panic!(
+                "get_players(): Lottery on: {}  players.is_empty(): {}",
+                self.lottery_is_on(),
+                self.players.is_empty()
+            );
+        }
+    }
+
+    // Random number generation from block timestamp
+    fn get_random_number(&mut self) -> u32 {
+        let timestamp: u64 = exec::block_timestamp();
+        let code_hash: sp_core::H256 = blake2_256(&timestamp.to_be_bytes()).into();
+        let u_buf = code_hash.to_fixed_bytes();
+        let mut number: u32 = 0;
+
+        for u in u_buf {
+            number += u as u32;
+        }
+
+        number
+    }
+
+    /// Lottery winner calculation
+    /// Requirements:
+    /// * Only owner can pick the winner
+    /// * Lottery has started and lottery time is expired
+    /// * List of players must not be empty
+    async fn pick_winner(&mut self) {
+        if msg::source() == self.lottery_owner
+            && self.lottery_state.lottery_started
+            && self.lottery_state.lottery_start_time + self.lottery_state.lottery_duration
+                <= exec::block_timestamp()
+            && !self.players.is_empty()
+        {
+            let index = (self.get_random_number() % (self.players.len() as u32)) as usize;
+            let win_player_index = *self.players.keys().nth(index).expect("Player not found");
+            let player = self.players[&win_player_index];
+
+            if self.token_address.is_some() {
+                debug!("Transfer tokens to the winner");
+                self.transfer_tokens(&exec::program_id(), &player.player_id, self.lottery_balance)
+                    .await;
+
+                self.lottery_balance = 0;
+            } else {
+                msg::send_bytes(player.player_id, b"Winner", exec::value_available());
+            }
+
+            self.lottery_history
+                .insert(self.lottery_id, player.player_id);
+            msg::reply(Event::Winner(win_player_index), 0);
+
+            debug!(
+                "Winner: {} token_address(): {:?}",
+                index, self.token_address
+            );
+
+            self.token_address = None;
+            self.lottery_state = LotteryState::default();
+            self.players = BTreeMap::new();
+        } else {
+            panic!(
+                "pick_winner(): Owner message: {}  lottery_duration: {}  players.is_empty(): {}",
+                msg::source() == self.lottery_owner,
+                self.lottery_state.lottery_start_time + self.lottery_state.lottery_duration
+                    > exec::block_timestamp(),
+                self.players.is_empty()
+            );
+        }
+    }
+}
+
+static mut LOTTERY: Option<Lottery> = None;
+
+#[gstd::async_main]
+async unsafe fn main() {
```

Comment:
No need to point `unsafe` as it is inside the macro.
```suggestion
async fn main() {
```